feat: initial setup for Sauce Visual (#2486)
saucelabs/sauce-docs: docs/visual-testing/integrations/cypress.md

---
sidebar_label: Cypress
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Cypress Integration

## Introduction

This guide requires an existing Cypress project.<br />
You can alternatively take a look to our [example repository](#example).

Sauce Visual provides an integration with [Cypress](https://cypress.io) through a plugin that you can add to any existing Cypress project.

Sauce Labs Visual introduce a new Cypress command:
- `cy.visualCheck()`: Takes a screenshot and send it to Sauce Labs Visual for comparison.

## Quickstart

### Step 1: Add Sauce Visual for Cypress dependency

> - Install the Sauce Visual for Cypress plugin in your current project.

Reviewer: Since we rely on Cypress, that means that Cypress won't be supported where our plugin can. 
Eventually we can mention which version is supported (currently at least 12+), not sure about node version.

> - Install the Sauce Visual for Cypress plugin in your current project.

Author: I've added more information on how to use Cypress on Sauce. That will guide the customers to a doc with what is supported

```sh

> npm install --save @saucelabs/cypress-visual-plugin

Reviewer: This depends a lot on user's setup. For those who have a Cypress project in a bigger project sure. For those who have a dedicated project for Cypress, it may not be effective in their CI context.

Both options are ok, IMO that's something users should be able to figure out.

> npm install --save @saucelabs/cypress-visual-plugin

Author: Changed it to `--save-dev`. Most customers would add this to an existing project as a dev dependency

> npm install --save @saucelabs/cypress-visual-plugin

Author: Changed is for all JS bindings to `--save-dev` for consistency

```

### Step 2: Configure Cypress to use Sauce Visual for Cypress plugin

- Import the plugin in Cypress project configuration, at the top of the file:
```ts
import { CypressSauceVisual } from '@saucelabs/cypress-visual-plugin';
```

- Register the plugin to Cypress events in Cypress project configuration:

``` ts
export default defineConfig({
  e2e: {
    [...]
    setupNodeEvents(on, config) {
      CypressSauceVisual.register(on, config);
    },
  },
})
```

- Register Sauce Visual for Cypress commands. Add the following line in your `cypress/support/e2e.ts`:
```ts
import '@saucelabs/cypress-visual-plugin/commands';
```

### Step 3: Add visual tests in your project:

```ts
context('Sauce Demo', () => {
  it('should render correctly', () => {
    ...
    cy.visualCheck('visual: my-homepage');
    ...
  })
});
```

### Step 4: Configure your Sauce Labs credentials

> Sauce Visual relies on environment variables for authentications.<br />

Reviewer: That's probably not the method we would suggest a user to use (even if there is already `user`/`key` supported in Cypress config. I would rather vote for keeping it this way, and update if we feel a need for it.

> Sauce Visual relies on environment variables for authentications.<br />

Author: Agree with @FriggaHel

> Both `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` need to be set prior starting your WebdriverIO job.

Author: @FriggaHel , copy-paste I think 😅 
```suggestion
Both `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` need to be set prior starting your Cypress job.
```

Reviewer: Correct ! Thanks !

Username and Access Key can be retrieved from https://app.saucelabs.com/user-settings.

```sh
export SAUCE_USERNAME=__YOUR_SAUCE_USER_NAME__
export SAUCE_ACCESS_KEY=__YOUR_SAUCE_ACCESS_KEY__
```

### Step 5: Run the test

Everything is now configured, your tests can be run as any other Cypress project.

> Builds will appear on Sauce Labs platform as soon as they have been created by the running tests: https://app.saucelabs.com/visual/builds.

Author: Added that

## Advanced usage

### Build name

Sauce Visual for Cypress plugin extends Cypress configuration, allowing to define the context, thus acting on which baselines new snapshots will be compared to. ([More info on baseline matching](../sauce-visual.md#baseline-matching))

Options:
- `region`: Sauce Labs Region where the new build will be created (default: `us-west-1`)
- `buildName`: Name of the build (default: `Cypress Visual Testing`)
- `project`: Name of the project (default: `None`)
- `branch`: Name of branch (default: `None`)

They need to be set through the `saucelabs` attribute of `e2e` configuration.

Example:
```javascript
export default defineConfig({
  e2e: {
    [...]
    saucelabs: {
      region: 'eu-central-1',
      buildName: 'SauceDemo - Cypress',
      project: 'Cypress examples',
      branch: 'main',
    },
    [...]
  }
}
```

### Ignored regions

In the case you need to ignore some region when running your tests, Sauce Visual provides a way to ignore user-specified areas.

Those ignored regions are specified when requesting a new snapshot.

> #### User-specified ignored region

Author: Please always start with the component based ignore regions (that's the best practice) and the the coordinate

Remember, the baseline is established during the initial run, and any subsequent visual differences detected will be marked for review.

## Advanced usage

### Build name

Sauce Visual for Cypress plugin extends Cypress configuration, allowing to define the context, thus acting on which baselines new snapshots will be compared to. ([More info on baseline matching](/visual-testing/#baseline-matching))

Options:

- `region`: Sauce Labs Region where the new build will be created (default: `us-west-1`)
- `buildName`: Name of the build (default: `Cypress Visual Testing`)
- `project`: Name of the project (default: `None`)
- `branch`: Name of branch (default: `None`)

They need to be set through the `saucelabs` attribute of `e2e` configuration.

Example:

```javascript
export default defineConfig({
  e2e: {
    [...]
    saucelabs: {
      region: 'eu-central-1',
      buildName: 'SauceDemo - Cypress',
      project: 'Cypress examples',
      branch: 'main',
    },
    [...]
  }
}
```

### Ignored regions

#### Component-based ignored region

> In case you need to ignore some components when running your tests, Sauce Labs Visual provides a way to ignore a list of components.

Reviewer 2: ```suggestion
Sauce Labs Visual provides a way to ignore a list of components.
```

An ignored component can be a specific element from the page.

Those ignored components are specified when requesting a new snapshot.

Example:

```javascript
    cy.sauceVisualCheck('login-page', {
      ignoredRegions: [
        cy.get('[data-test="username"]'),
      ]
    });
```

#### User-specified ignored region

Alternatively, ignored regions can be user-specified areas. A region is defined by four elements.

- `x`, `y`: The location of the top-left corner of the ignored region
- `width`: The width of the region to ignore
- `height`: The heigh of the region to ignore

_Note: all values are pixels_

Example:

```javascript
    cy.sauceVisualCheck('login-page', {
      ignoredRegions: [
        {
          x: 240,
          y: 800,
          width: 1520,
          height: 408
        }
      ],
    });
```

### Specifying options for Cypress Screenshot

> Sauce Visual is relying on native screenshot feature from Cypress. As `cy.snapshot` has options, they can be specified in the `cy.visualCheck` command.

Reviewer 2: As `cy.snapshot` has options, they can be specified in the `cy.visualCheck` command.

I'm not fully sure what we want to say here
```suggestion
Sauce Visual relies on the native screenshot feature from Cypress. Options in cy.snapshot can be specified in the cy.visualCheck command.
```
